networks/2b_community_detection.py

```diff
@@ -0,0 +1,202 @@
+#     display_name: Python [conda env:core_acc] *
+#     language: python
+#     name: conda-env-core_acc-py
+# ---
+
+# # Network community detection
+#
+# This notebook performs community detection approaches to identify network modules.
+#
+# Note: All methods here are using undirected weighted networks. All methods take edge weights as input.
+
+# +
+import os
+import random
+
+import numpy as np
+import pandas as pd
+import igraph as ig
+from core_acc_modules import paths
+
+# +
+# User params
+
+# Choices = ["fastgreedy", "walktrap", "louvain", "infomap"]
+method = "infomap"
+
+# Params for different methods to adjust
+# length of random walk to perform for walktrap
+# Short random walks tend to stay in the same community
+nsteps = 10
+
+# Number of trials to attempt to partition the network for infomap
+ntrials = 20
+
+# TO DO
+# Think about running these methods across multiple seeds and taking a consensus
+
+# +
+# Load correlation matrix
+pao1_pearson_mat_filename = paths.PAO1_CORR_LOG_SPELL
+pa14_pearson_mat_filename = paths.PA14_CORR_LOG_SPELL
+
+# Take abs of correlation scores
+# In this case we care about the strength and not the direction
+pao1_corr = pd.read_csv(
+    pao1_pearson_mat_filename, sep="\t", index_col=0, header=0
+).abs()
+pa14_corr = pd.read_csv(
+    pa14_pearson_mat_filename, sep="\t", index_col=0, header=0
+).abs()
+# -
+
+pao1_corr.head()
+
+# +
+# Format correlation matrix into graph (i.e. dataframe with edge weight per pair of genes)
+# The dataframe should have columns: from, to, weight
+pao1_corr_graph = pao1_corr.stack().reset_index()
+pao1_corr_graph.columns = ["from", "to", "weight"]
+
+pa14_corr_graph = pa14_corr.stack().reset_index()
+pa14_corr_graph.columns = ["from", "to", "weight"]
+# -
+
+# Drop duplicate rows since correlation matrix is symmetric
+pao1_corr_graph = pao1_corr_graph.drop_duplicates()
+pa14_corr_graph = pa14_corr_graph.drop_duplicates()
+
+# Drop gene loops
+# Note 'query' not working for some reason
+pao1_corr_graph = pao1_corr_graph[pao1_corr_graph["from"] != pao1_corr_graph["to"]]
+pa14_corr_graph = pa14_corr_graph[pa14_corr_graph["from"] != pa14_corr_graph["to"]]
+
+print(pao1_corr_graph.shape)
+pao1_corr_graph.head()
+
+# Make into a graph object
+pao1_G = ig.Graph.TupleList(pao1_corr_graph.values, weights=True, directed=False)
+pa14_G = ig.Graph.TupleList(pa14_corr_graph.values, weights=True, directed=False)
+
+# make sure vertex/edge properties exist
+print(pao1_G.es["weight"][:5])
+
+# ## Community detection
+
+# ### Fast-greedy
+# This algorithm starts from a completely unclustered set of nodes and iteratively adds communities such that the modularity (score maximizing within edges and minimizing between edges) is maximized until no additional improvement can be made.
+#
+# Note: Looks like fast-greedy requires a simple graph (i.e. no multiple edges per node), so we use [simplify](https://igraph.org/python/doc/api/igraph._igraph.GraphBase.html#simplify) to combine edges
```

Comment:
Hm, interesting. So one node/gene is connected only to a single node? This is what "simplify" does?

Reply:
This was my quick read on `simplify` though I was still not able to get fast-greedy to run. After spending some time to try to debug this, I just moved forward with the other methods since affinity seems to be giving us close to expected results